internal/safebuffer/safebuffer.go

```diff
@@ -0,0 +1,88 @@
+package safebuffer
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"sync"
+
+	"github.com/joshuarubin/teleport-job-worker/internal/safebuffer/safereader"
+)
+
+// Buffer is a goroutine safe buffer that ingests data as an io.Writer and
+// provides io.ReadClosers to reply and continue to stream incoming data.
+type Buffer struct {
+	mu      sync.RWMutex
+	buf     bytes.Buffer
+	jobDone <-chan struct{}
+	chans   map[chan []byte]<-chan struct{}
+}
+
+// ensure safeBuffer implements the io.Writer interface
+var _ io.Writer = (*Buffer)(nil)
+
+// New creates a new Buffer. Readers created with NewReader will be
+// automatically closed when the context becomes done or the jobDone channel
+// closes.
+func New(ctx context.Context, jobDone <-chan struct{}) *Buffer {
+	b := Buffer{
+		jobDone: jobDone,
+		chans:   map[chan []byte]<-chan struct{}{},
+	}
+	go b.jobWatcher(ctx)
+	return &b
+}
+
+// jobWatcher waits for the context to become done or for jobDone to close at
+// which point it will close all reader data channels and remove them from the
+// buffer's internal map.
+func (b *Buffer) jobWatcher(ctx context.Context) {
+	select {
+	case <-ctx.Done():
+	case <-b.jobDone:
+	}
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	for dataCh := range b.chans {
+		close(dataCh)
+		delete(b.chans, dataCh)
+	}
```

Comment:
I thought a lot about this concern, and tried to test it too. Based on what I understand from the docs and the results of my test, no, but I'm still not fully convinced either.

The [docs](https://pkg.go.dev/os/exec#Cmd.Wait) for `cmd.Wait()` say:
> Wait waits for the command to exit and waits for any copying to stdin or copying from stdout or stderr to complete.

Let's presume we're not at these lines because the `ctx` became done (in which case all bets are off). So in order to get to the loop you highlighted, the `jobDone` channel must have been closed. That channel, necessarily, can only be closed after `Wait()` completes. This means that all `Write()` calls to the `SafeBuffer`, which are synchronous, must complete before `dataCh` can be closed. The `SafeReader` may receive data via channel in a separate goroutine, which, once received unblocks the `Write()` call in the `SafeBuffer`. `Write()` holds the mutex for the entire time it executes, and so does `jobWatcher()` when it gets to that loop. This means that all of the channels will receive all of their data before `close()` is called on them and so all of the data will make it through the channel.